File: problems/019/019.c
#include <stdio.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    /* The first element is a dummy element. */
    const int MONTH_CONSTANT[13] = { 0, 1, -1, 0, 0, 1, 1, 2, 3, 3, 4, 4, 5 };

    int n = 0;

    for (int i = 1901; i <= 2000; i++)
    {
        for (int j = 1; j <= 12; j++)
        {
            int day_val = ( (365 * (i - 1))  + (int)((i - 1) / 400) 
                        + (int)((i - 1) / 4) - (int)((i - 1) / 100) 
                        + (30 * (j - 1)) + MONTH_CONSTANT[j-1] + 1  );

            if (((i % 4) == 0) && (j > 2))
                day_val++;

            day_val %= 7;

            if (day_val == 0)
                n++;
        }
    }

    printf("%d\n", n);

    return 0;
}

File: problems/019/019.f90
PROGRAM F019

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    ! The first element is a dummy element.
    INTEGER, PARAMETER, DIMENSION(13) :: MONTH_CONSTANT = &
    (/ 0, 1, -1, 0, 0, 1, 1, 2, 3, 3, 4, 4, 5 /)

    INTEGER :: N = 0

    INTEGER :: DAY_VAL, I, J

    DO I = 1901, 2000

        DO J = 1, 12

            DAY_VAL =         (((I - 1) * 365) + FLOOR(REAL(I - 1) / 400.0)    &
                    + FLOOR(REAL(I - 1) / 4.0) - FLOOR(REAL(I - 1) / 100.0)    &
                    +          ((J - 1) *  30) + MONTH_CONSTANT(J) + 1)

            IF (MOD(I, 4) == 0 .AND. J > 2) DAY_VAL = DAY_VAL + 1

            DAY_VAL = MOD(DAY_VAL, 7)

            IF (DAY_VAL == 0) N = N + 1

        END DO

    END DO

    PRINT ('(I0)'), N

END PROGRAM F019
